pkg/hive/hive.go

```diff
@@ -0,0 +1,297 @@
+	return h.envPrefix + upper
+}
+
+// registerCells registers the command-line flags from all the cells.
+func (h *Hive) registerFlags(parent *pflag.FlagSet) error {
+	for _, cell := range h.cells {
+		cell.registerFlags(h.flags)
+	}
+	var err error
+	h.flags.VisitAll(func(f *pflag.Flag) {
+		if err != nil {
+			return
+		}
+		if parent.Lookup(f.Name) != nil {
+			err = fmt.Errorf("Error registering flags: '%s' already registered", f.Name)
+		} else {
+			viper.BindEnv(f.Name, h.getEnvName(f.Name))
+			parent.AddFlag(f)
+		}
+	})
+	if err != nil {
+		return err
+	}
+	return h.viper.BindPFlags(h.flags)
+}
+
+// populate creates the cell configurations from viper and returns the combined
+// fx option for all cells and their configurations.
+func (h *Hive) populate(overrides ...any) (fx.Option, error) {
+	overridesMap := map[reflect.Type]any{}
+	for _, cfg := range overrides {
+		overridesMap[reflect.TypeOf(cfg)] = cfg
+	}
+
+	allOpts := []fx.Option{}
+	for _, cell := range h.cells {
+		cell := cell
+
+		cellOpts := cell.opts
+		if cell.newConfig != nil {
+			cfg := cell.newConfig()
+
+			if override, ok := overridesMap[reflect.TypeOf(cfg)]; ok {
+				cfg = override
+			} else {
+				if err := h.unmarshalConfig(&cfg); err != nil {
+					return nil, err
+				}
+			}
+			cellOpts = append(cellOpts, fx.Supply(cfg))
+		}
+
+		var cellOpt fx.Option
+		if cell.name != "" {
+			// Provide a logger to the cell that has subsys set to the cell name.
+			cellLogger := fx.Decorate(
+				func(log logrus.FieldLogger) logrus.FieldLogger {
+					return log.WithField(logfields.LogSubsys, cell.name)
+				})
+			cellOpts = append(cellOpts, cellLogger)
+			cellOpt = fx.Module(cell.name, cellOpts...)
+		} else {
+			cellOpt = fx.Options(cellOpts...)
+		}
+		allOpts = append(allOpts, cellOpt)
+	}
+	return fx.Options(allOpts...), nil
+}
+
+// createApp creates the fx application, unless already created.
+func (h *Hive) createApp() error {
+	if h.app != nil {
+		return nil
+	}
+	opts, err := h.populate()
+	if err != nil {
+		return err
+	}
+	h.app = fx.New(
+		fx.WithLogger(func() fxevent.Logger { return h.fxLogger }),
+		fx.Supply(fx.Annotate(log, fx.As(new(logrus.FieldLogger)))),
+		fx.Decorate(func(parent fx.Lifecycle) fx.Lifecycle {
+			h.lifecycle.parent = parent
+			return &h.lifecycle
+		}),
+		fx.StartTimeout(h.startTimeout),
+		fx.StopTimeout(h.stopTimeout),
+		fx.Populate(&h.dotGraph),
+		opts,
+	)
+	return h.app.Err()
+}
+
+// TestApp constructs a test application for the hive with given configuration
+// overrides.
+func (h *Hive) TestApp(tb fxtest.TB, configs ...any) (*fxtest.App, error) {
+	opts, err := h.populate(configs...)
+	if err != nil {
+		return nil, err
+	}
+	return fxtest.New(tb,
+		fx.Supply(fx.Annotate(log, fx.As(new(logrus.FieldLogger)))),
+		opts), nil
+}
+
+func (h *Hive) unmarshalConfig(target *any) error {
+	if err := h.viper.Unmarshal(target, decoderConfig); err != nil {
+		return fmt.Errorf("Failed to unmarshal config struct %T: %w.\n"+
```

Comment:
nit: capitalized error string (`error strings should not be capitalized (ST1005) go-staticcheck`)

Reply:
Fixed.